src/bun.js/child_process.exports.js

```diff
@@ -1,3 +1,995 @@
+
+function validateBoolean(value, name) {
+  if (typeof value !== "boolean")
+    throw new ERR_INVALID_ARG_TYPE(name, "boolean", value);
+}
+
+/**
+ * @callback validateOneOf
+ * @template T
+ * @param {T} value
+ * @param {string} name
+ * @param {T[]} oneOf
+ */
+
+/** @type {validateOneOf} */
+const validateOneOf = (value, name, oneOf) => {
+  // const validateOneOf = hideStackFrames((value, name, oneOf) => {
+  if (!ArrayPrototypeIncludes(oneOf, value)) {
+    const allowed = ArrayPrototypeJoin(
+      ArrayPrototypeMap(oneOf, (v) =>
+        typeof v === "string" ? `'${v}'` : String(v)
+      ),
+      ", "
+    );
+    const reason = "must be one of: " + allowed;
+    throw new ERR_INVALID_ARG_VALUE(name, value, reason);
+  }
+};
+
+/**
+ * @callback validateObject
+ * @param {*} value
+ * @param {string} name
+ * @param {{
+ *   allowArray?: boolean,
+ *   allowFunction?: boolean,
+ *   nullable?: boolean
+ * }} [options]
+ */
+
+/** @type {validateObject} */
+const validateObject = (value, name, options = null) => {
+  // const validateObject = hideStackFrames((value, name, options = null) => {
+  const allowArray = getOwnPropertyValueOrDefault(options, "allowArray", false);
+  const allowFunction = getOwnPropertyValueOrDefault(
+    options,
+    "allowFunction",
+    false
+  );
+  const nullable = getOwnPropertyValueOrDefault(options, "nullable", false);
+  if (
+    (!nullable && value === null) ||
+    (!allowArray && ArrayIsArray(value)) ||
+    (typeof value !== "object" &&
+      (!allowFunction || typeof value !== "function"))
+  ) {
+    throw new ERR_INVALID_ARG_TYPE(name, "Object", value);
+  }
+};
+
+/**
+ * @callback validateArray
+ * @param {*} value
+ * @param {string} name
+ * @param {number} [minLength]
+ * @returns {asserts value is any[]}
+ */
+
+/** @type {validateArray} */
+const validateArray = (value, name, minLength = 0) => {
+  // const validateArray = hideStackFrames((value, name, minLength = 0) => {
+  if (!ArrayIsArray(value)) {
+    throw new ERR_INVALID_ARG_TYPE(name, "Array", value);
+  }
+  if (value.length < minLength) {
+    const reason = `must be longer than ${minLength}`;
+    throw new ERR_INVALID_ARG_VALUE(name, value, reason);
+  }
+};
+
+/**
+ * @callback validateString
+ * @param {*} value
+ * @param {string} name
+ * @returns {asserts value is string}
+ */
+
+/** @type {validateString} */
+function validateString(value, name) {
+  if (typeof value !== "string")
+    throw new ERR_INVALID_ARG_TYPE(name, "string", value);
+}
+
+/**
+ * @param {?object} options
+ * @param {string} key
+ * @param {boolean} defaultValue
+ * @returns {boolean}
+ */
+function getOwnPropertyValueOrDefault(options, key, defaultValue) {
+  return options == null || !ObjectPrototypeHasOwnProperty(options, key)
+    ? defaultValue
+    : options[key];
+}
+
+function ObjectPrototypeHasOwnProperty(obj, prop) {
+  return Object.prototype.hasOwnProperty.call(obj, prop);
+}
+
+function ArrayPrototypePush(array, ...items) {
```

Comment:
Node uses wrappers for builtin functions like these to avoid issues caused by userland code overriding functions on prototypes. They call these "primordials".

For now, in Bun, we should explicitly set this to `var ArrayPrototypePush = Array.prototype.push` at the top of the file and then `.call` or `.apply` instead of properties on the Array object. It will be simpler to refactor later.

Longer-term we will need to move these to use builtins we pass to modules like these to avoid any issues with users overriding globals before this module is evaluated

Reply:
Yeah that makes sense. I knew there was a more correct way to simulate primordials. I'll refactor this tonight as I'm wrapping things up